os/bluestore: handle spurious read errors

emmericp opened the pull request:
Some kernels (4.9+) sometime fail to return data when reading
from a block device under memory pressure. This patch retries
the read if the checksum verification fails, tests show that
the first retried read succeeds in ~99.5% of the cases, so
3 attempts are made before giving up on the data.

Works-around: http://tracker.ceph.com/issues/22464
Signed-off-by: Paul Emmerich <paul.emmerich@croit.io>

liewegas commented:
"Fixes" might be overstating it about.. "works around" perhaps!

This seems like a fine workaround.  The primary gap is that it doesn't cover the rocksdb case where rocksdb's read gets the zeros and not bluestores.  That would require a similar hack in rocksdb, though.

Has there been any success narrowing this down to a specific range of kernels?  Mark was having trouble reproducing it so we stopped the hunt.. but if you have a reproducer it would be great to at least verify that the newer mainline kernels are no longer affected.  Or if they are, we can get the MM folks involved to fix the bug!

emmericp commented:
I (luckily) never saw the rocksdb variant of it.

I've seen it on our systems with 4.9 (debian stable) and on 4.13, 4.14, and 4.16 in debian backports. I've also seen it with a centos ml kernel from epel but I don't recall the specific version.

I unfortunately cannot reproduce it in a reliable way on our test servers, it takes a few attempts and the process is somewhat annoying.
I've got a production cluster where this happens every few days on some OSDs, but I'd rather not play around with kernel versions there. But I verified that retrying the read almost always works there.

I'm going to test 4.17 as soon as it lands in stretch-backports.

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -7469,7 +7470,11 @@ int BlueStore::_do_read(
 		       b2r_it->second.front().logical_offset) < 0) {
-	return -EIO;
+        if (retries) {
+          return _do_read(c, o, offset, length, bl, op_flags, retries - 1);
```
Suggest to log here and there with level level small enough to see that by default.

ifed01 replied:
Not 100% sure but may be it makes sense to retry in case of zero-filled block only.
And it would be nice to have a comment (referring to a ticket) why we're doing all that stuff here.

emmericp replied:
* _verify_csum already logs to derr if the verification fails - or did you mean to log if a retried read succeeds?

* I *think* that we might get reads that are only partially zero (cf. the rocksdb version of that issue)

* Comments: sure, will add that

liewegas replied:
You're right, pages will be fully zeroed (at least that's what we've seen) but the crc may be over multiple pages.  Comments++!

ifed01 replied:
and yet another suggestion - let's make this retry count configurable.

xiexingguo commented:
This is probably not enough, I've encountered another similar issue while trying to load onode from RocksDB yesterday :
```
abort: Corruption: block checksum mismatch

 ceph version 12.2.2-16-50-g423925c (423925c8e8691e59ec35f4c568a76d03760f3d0c) luminous (stable)
 1: (()+0xa599a1) [0x7f87f77789a1]
 2: (()+0xf5e0) [0x7f87f49d65e0]
 3: (gsignal()+0x37) [0x7f87f39fe1f7]
 4: (abort()+0x148) [0x7f87f39ff8e8]
 5: (RocksDBStore::get(std::string const&, char const*, unsigned long, ceph::buffer::list*)+0x1ce) [0x7f87f76d084e]
 6: (BlueStore::Collection::get_onode(ghobject_t const&, bool)+0x545) [0x7f87f767b865]
 7: (BlueStore::getattr(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, ghobject_t const&, char const*, ceph::buffer::ptr&)+0
x108) [0x7f87f767e368]
 8: (PGBackend::objects_get_attr(hobject_t const&, std::string const&, ceph::buffer::list*)+0xc7) [0x7f87f7427887]
 9: (PrimaryLogPG::get_object_context(hobject_t const&, bool, std::map<std::string, ceph::buffer::list, std::less<std::string>, s
td::allocator<std::pair<std::string const, ceph::buffer::list> > > const*)+0x9d7) [0x7f87f73a2417]
 10: (PrimaryLogPG::find_object_context(hobject_t const&, std::shared_ptr<ObjectContext>*, bool, bool, hobject_t*)+0x323) [0x7f87
f73a5c53]
 11: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x21e4) [0x7f87f73dcd54]
 12: (PrimaryLogPG::do_request(boost::intrusive_ptr<OpRequest>&, ThreadPool::TPHandle&)+0xebb) [0x7f87f739dcfb]
 13: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x42e) [0x7f87f71ff7ce]
 14: (PGQueueable::RunVis::operator()(boost::intrusive_ptr<OpRequest> const&)+0x57) [0x7f87f74a02f7]
 15: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x1282) [0x7f87f7232d32]
 16: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x839) [0x7f87f77bd779]
 17: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x7f87f77bf710]
 18: (()+0x7e25) [0x7f87f49cee25]
 19: (clone()+0x6d) [0x7f87f3ac134d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

```
Is that related?

liewegas commented:
@xiexingguo what kernel version?

xiexingguo commented:
> @xiexingguo what kernel version?

3.10.0-327

xiexingguo commented:
It's a quite old and stable kernel, so it is probably not related? I'll check if I can find a reproducer first...

emmericp commented:
updated:

* comments
* `dout(5)` if a read was successful after a retry (errors themselves are already logged by `_verify_csum`)
* it's now a configurable option: `bluestore_retry_disk_reads` (any preferences on the name?)


@xiexingguo: yours looks related (but different) to http://tracker.ceph.com/issues/22102 which I unfortunately can't reproduce :(

emmericp commented:
@ifed01 or @liewegas could one of you please review my updated commit?

xiexingguo commented:
The fix might never??? get a change to be exercised. Should we introduce some failure injection mechanism or something similar?

liewegas commented:
Yeah, that would be good.. bdev_inject_read_err_probability or something?

The problem, though, is that we aren't covering the rocksdb read case... and that ones a bit more work since the checksum is verified inside rocksdb, so the retry would have to be implemented there.

emmericp commented:
I've added bdev_inject_crc_err_probability

ifed01 commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -7257,7 +7257,8 @@ int BlueStore::_do_read(
   bufferlist& bl,
-  uint32_t op_flags)
+  uint32_t op_flags,
+  uint8_t retry_count)
```
IMO no much sense in using that narrow int type here, size_t should be fine.

emmericp replied:
I disagree; an actual real checksum error will trigger this always. Really high retry counts will at least generate lots of IOPS or hang the thread. I guess users could guess that lots of retries are always good, so I've restricted the option to 255 specificly.

Or would you just like to change the variable type here without adjusting the value range allowed by the option?

liewegas replied:
Right, I think he's just talking about the variable type.

ifed01 commented:
OK, we have inject_probability now but the code is still never executed... IMO worth to have test coverage in store_test. E.g. do massive reads with pretty high error probability and retry count. 
Will probably need some tricks to ensure all reads are finally 100% successful though... To ensure test case has 100% predictable behavior.

Also it might make sense to introduce a performance counter to track such retries which is both useful in production and can be checked while testing.

liewegas commented:
I agree a perfcounter would be nice. The problem with adding a test at this point is that we can't inject at the bdev level without also triggering crc errors for rocksdb, which aren't handled seamlessly.

liewegas commented:
FWIW I don't think any of these are blocking issues (but the perfcounter would be nice :).

Does anybody have any thoughts on how to tackle the rocksdb crc mismatch handling?

emmericp commented:
I've added a test to store_test.cc; the chances of it randomly going wrong are basically zero, see comments in code for a derivation.

@liewegas I don't see a problem for rocksdb with my test, am I missing something?

Not sure if a performance counter is worth the effort; it already logs an error with level -1 which should show up in syslog and alert someone/thing? (But it might be a nice Ceph programming exercise for me, so I might add one if I find some time...)

liewegas commented:
The perfcounter means that it can be monitored across the cluster via prometheus or whatever.

rocksdb issues reads via bluefs in the BlockDevice layer, which may inject a zeroed page.  If that happens, it will trigger a crc mismatch inside rocksdb.  You probably won't hit it in a small-scale test because everything will be in rocksdb's cache and it won't be reading anything.  It should trigger under a larger scale test (probably anything that triggers compaction if the probability is high enough). The backtrace would be something like the one in bug http://tracker.ceph.com/issues/22102

emmericp commented:
I've added a perf counter and changed the retry_count variable type to match the type of the option (which is still restricted to 0 to 255).

The error injection in `Bluestore::_verify_csum`, not in the actual reading, so rocksdb isn't affected -- _verify_csum is only called in two locations in Bluestore, both of which have the retry logic.
But that just means I named the debug inject option poorly; I've renamed it to `bluestore_debug_inject_csum_err_probability` to reflect what it actually does

liewegas commented:
Perfect!  Thanks

tchaikov commented:
- http://pulpito.ceph.com/kchai-2018-09-10_07:18:46-rados-wip-kefu-testing-2018-09-10-1219-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-11_07:48:23-rados-wip-kefu-testing-2018-09-10-1219-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-11_09:51:05-rados-wip-kefu-testing-2018-09-10-1219-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-11_11:46:31-rados-wip-kefu-testing-2018-09-10-1219-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-11_12:12:46-rados-wip-kefu-testing-2018-09-10-1219-distro-basic-mira/

ifed01 commented:
@liewegas 
> Does anybody have any thoughts on how to tackle the rocksdb crc mismatch handling?
Just to recall - the outcome of the read error here is that we get some blocks (4K pages) all set to zero, right? 
If so we can inspect read data at BlockDevice level for zero pages and reread them similar to this PR (which is configurable behavior, disabled by default. May be provided for bluefs reads only). Some overhead but it's to be enabled on "broken" deployments only.
The next step might be to implement a sort of dedup to avoid zero pages saving at all hence any such read is definitely an error...

TheJJ commented:
Please backport! Except the test function, the patch applies against 13.2.2 :smiley:.

Without this patch, we had many unnecessary deep scrubs. This patch fixes it.

TheJJ commented:
Backport patch is here: https://github.com/croit/ceph/commit/35cb7d9afb5486ad4d9096a8984ec63f789dd5c8.

janpekar commented:
Is it planned to get this "fix" into luminous release?

liewegas commented:
yeah, we should include this in mimic and luminous both.

liewegas commented:
https://github.com/ceph/ceph/pull/24649
https://github.com/ceph/ceph/pull/24647

Osa1989 commented:
I am getting these errors on Ubuntu 18.04 with Kernel 4.15.0-42-generic and ceph version 13.2.2
How can I apply the patch?

emmericp commented:
> You may got me wrong. I know that THIS is not a fix - but this was not what i asked for.

I'm not aware of any changes or fixes in the kernel that would mitigate this

disaster123 commented:
@emmericp OK than this was just by accident. What i'm still wondering is, whether this is a kernel bug or feature? If ceph needs a workaround a would consider it a bug. But beeing present since 4.9 makes me think the kernel devs think this is a feature?

liewegas commented:
We're pretty certain this is (was?) a kernel bug, but we never succeeded in figuring out what [patch] caused it or what fixed it.  We only know that the symptoms appear to go away with newer kernels, as you noticed.  :/  Not very satisfying, I know, but that's all we know!

disaster123 commented:
Is there any open bug report at kernel.org or on the ML?
